### PR TITLE
Initial work on Multiplayer support for FS19

### PR DESCRIPTION
**Pull request by davboecki**
Some initial work I did on multiplayer support for courseplay FS19.
This is definitely not ready to be merged.

This is really rough. 
Some design decisions may need rework. 

What does work: 
* Basic Drive functionality
* Course recording
* Course saving and loading
* Probably more

What works now after more involve fixing:
* AutoDrive course usage in MP
* The course generation GUI
* The Advanced Config GUI.

What definitely does not work:
* Probably more

If people want to test this, go ahead. It is playable, at least from my initial tests. 
Feedback and major problems/bug reports welcome.

### CODE REVIEW

File: course_management.lua [hunk 1]
```diff
@@ -302,7 +306,11 @@ function courseplay:loadCourse(vehicle, id, useRealId, addCourseAtEnd) -- fn is
 
 		-- SETUP 2D COURSE DRAW DATA
 		vehicle.cp.course2dUpdateDrawData = true;
-		courseplay.hud:setReloadPageOrder(vehicle, vehicle.cp.hud.currentPage, true)
+		courseplay.hud:setReloadPageOrder(vehicle, vehicle.cp.hud.currentPage, vehicle.Waypoints)
```
Comment (davboecki) [inline course_management.lua]: Node to self: What is this change? (Probably wrong)

File: register.lua [hunk 2]
```diff
@@ -17,6 +17,8 @@ function courseplay.registerEventListeners(vehicleType)
 	SpecializationUtil.registerEventListener(vehicleType, "onDelete", courseplay)
 	SpecializationUtil.registerEventListener(vehicleType, "onRegisterActionEvents", courseplay)
 	SpecializationUtil.registerEventListener(vehicleType, "onPostDetachImplement", courseplay)
+    SpecializationUtil.registerEventListener(vehicleType, "onReadStream", courseplay)
```
Comment (davboecki) [inline register.lua]: Formatting

File: settings.lua [hunk 3]
```diff
@@ -264,6 +278,7 @@ function courseplay:switchCourseplayerSide(combine)
 		else
 			combine.cp.forcedSide = nil;
 		end;
+		combine:setCpVar('stopAtEnd', combine.cp.stopAtEnd, courseplay.isClient);
```
Comment (pvaiko) [inline settings.lua]: FYI, all these legacy settings will at some point be converted to a SettingsList or derived object. Those (at the end of settings.lua) should probably have the MP updates built in later.

Comment (davboecki) [reply settings.lua]: Good to know, but all of there changes are probably going away anyway. They were just a result of me having huge desync issues, caused by other things, which are mostly fixed already.

Comment (pvaiko) [reply settings.lua]: All the advanced config settings are SettingsList and derived classes so if a sync of those needed, probably that would be the place to implement it.

**Comment by davboecki** [comment]
OK, the advanced GUI Stuff and Settings Syncing are now fixed/implemented. 
I'm not aware of any more bugs at the moment. But I haven't done any intensive testing, so I still need feedback from others who may have tested this.

**Comment by pvaiko** [comment]
Awesome, I have no problem merging this and push a new release if you are ready to fix the issues reported.

Not so worried about problems with MP, as long as the single player works I'm fine.

**Comment by davboecki** [comment]
Nearly all changes are MP related. I have a clean log when I'm testing and I never had any issues in SP (Apart from trying to use mode 2...) only MP all the way. So I'm fine with merging this and then fixing the leftover issues. 

Just as a Side Node I tried to run this with the CombineUnloadAIDriver branch. (see https://github.com/davboecki/courseplay/tree/multiplayerCombineUnloadAIDriver) SP appears to work but MP does not. So best not to merge that up there until I have figured out, what is wrong there.

**Comment by davboecki** [comment]
Hold on there. I may have found the Issue with the CombineUnloadAIDriver branch and It might happen in master as well. Will push a fix to this and report back.

**Comment by davboecki** [comment]
OK, that is fixed now as well. 
It should be ready to merge now.

**Pull request merged**